diff -u at-3.1.8/Makefile.in.orig at-3.1.8/Makefile.in
--- at-3.1.8/Makefile.in.orig	Wed Feb 14 13:59:09 2001
+++ at-3.1.8/Makefile.in	Mon Feb 19 14:48:42 2001
@@ -51,6 +51,8 @@
 
 OTHERS		= parsetime.l parsetime.y
 
+TEST_VERBOSE	= 0
+
 DOCS =  Problems Copyright README ChangeLog timespec
 
 MISC =  COPYING  Makefile.in configure acconfig.h install-sh \
@@ -150,6 +152,9 @@
 
 parsetest: lex.yy.c y.tab.c
 	$(CC) -o parsetest $(CFLAGS) $(DEFS) -DTEST_PARSER -DNEED_YYWRAP lex.yy.c y.tab.c
+
+test: parsetest
+	PERL_DL_NONLAZY=1 perl -e 'use Test::Harness qw(&runtests $$verbose); $$verbose=$(TEST_VERBOSE); runtests @ARGV;' test.pl
 
 .depend: $(CSRCS)
 	gcc $(CFLAGS) $(DEFS) -MM $(CSRCS) > .depend
diff -u at-3.1.8/test.pl.orig at-3.1.8/test.pl
