--- at-3.1.10/at.c.perm	2007-07-02 15:40:35.000000000 +0200
+++ at-3.1.10/at.c	2007-07-02 16:19:54.000000000 +0200
@@ -314,10 +314,11 @@
 	 * bit.  Yes, this is a kluge.
 	 */
 	cmask = umask(S_IRUSR | S_IWUSR | S_IXUSR);
-        seteuid(real_uid);
+        //seteuid(real_uid);
+	seteuid(effective_uid);
 	if ((fd = open(atfile, O_CREAT | O_EXCL | O_TRUNC | O_WRONLY, S_IRUSR)) == -1)
 	    perr("Cannot create atjob file %.500s", atfile);
-        seteuid(effective_uid);
+        //seteuid(effective_uid);
 
 	if ((fd2 = dup(fd)) < 0)
 	    perr("Error in dup() of job file");
