--- at-3.1.10/Makefile.in.debug	2007-03-06 11:35:28.000000000 +0100
+++ at-3.1.10/Makefile.in	2007-03-06 14:50:07.000000000 +0100
@@ -104,14 +104,14 @@
 	chown $(DAEMON_USERNAME):$(DAEMON_GROUPNAME) $(IROOT)$(ATJOB_DIR) $(IROOT)$(ATSPOOL_DIR)
 	test -f $(IROOT)$(etcdir)/at.allow || test -f $(IROOT)$(etcdir)/at.deny || $(INSTALL) -m 600 at.deny $(IROOT)$(etcdir)/
 	$(INSTALL) -o $(INSTALL_ROOT_USER) -g $(DAEMON_GROUPNAME) pam_atd $(IROOT)$(etcdir)/pam.d/atd
-	$(INSTALL) -m 4755 -s at $(IROOT)$(bindir)
+	$(INSTALL) -m 4755 at $(IROOT)$(bindir)
 	$(LN_S) -f at $(IROOT)$(bindir)/atq
 	$(LN_S) -f at $(IROOT)$(bindir)/atrm
 	$(INSTALL) -m 755 batch $(IROOT)$(bindir)
 	$(INSTALL) -m 755 -d $(IROOT)$(man1dir)
 	$(INSTALL) -m 755 -d $(IROOT)$(man5dir)
 	$(INSTALL) -m 755 -d $(IROOT)$(man8dir)
-	$(INSTALL) -m 755 -s atd $(IROOT)$(sbindir)
+	$(INSTALL) -m 755 atd $(IROOT)$(sbindir)
 	$(INSTALL) -m 755 atrun $(IROOT)$(sbindir)
 	$(INSTALL) -m 644 at.1 $(IROOT)$(man1dir)/
 	cd $(IROOT)$(man1dir) && $(LN_S) -f at.1 atq.1 && $(LN_S) -f at.1 batch.1 && $(LN_S) -f at.1 atrm.1
