--- at-3.1.10/pam_atd.newpam	2007-01-30 13:04:19.000000000 +0100
+++ at-3.1.10/pam_atd	2007-01-30 13:02:09.000000000 +0100
@@ -0,0 +1,15 @@
+# The PAM configuration file for the at daemon
+#
+#
+auth		sufficient	pam_rootok.so
+auth		required	pam_stack.so 
+auth		include		system-auth
+auth		required	pam_env.so
+account		required	pam_stack.so 
+account		include		system-auth
+session		required	pam_stack.so 
+session		include		system-auth
+# Sets up user limits, please uncomment and read /etc/security/limits.conf
+# to enable this functionality.
+# session    required   pam_limits.so
+#
