--- at-3.1.8/Makefile.in.pie	2004-05-12 11:11:07.690785433 +0200
+++ at-3.1.8/Makefile.in	2004-05-12 11:23:10.367957697 +0200
@@ -69,13 +69,13 @@
 all: at atd atrun
 
 at: $(ATOBJECTS)
-	$(CC) $(CFLAGS) -o at $(ATOBJECTS) $(LIBS) $(LEXLIB)
+	$(CC) $(CFLAGS) -o at -pie $(ATOBJECTS) $(LIBS) $(LEXLIB)
 	rm -f $(CLONES)
 	$(LN_S) -f at atq
 	$(LN_S) -f at atrm
 
 atd: $(RUNOBJECTS)
-	$(CC) $(CFLAGS) -o atd $(RUNOBJECTS) $(LIBS) $(SELINUXLIB)
+	$(CC) $(CFLAGS) -o atd -pie $(RUNOBJECTS) $(LIBS) $(SELINUXLIB)
 
 y.tab.c y.tab.h: parsetime.y
 	$(YACC) -d parsetime.y
@@ -87,7 +87,7 @@
 	configure
 
 .c.o:
-	$(CC) -c $(CFLAGS) $(DEFS) $*.c
+	$(CC) -c $(CFLAGS) -fpie $(DEFS) $*.c
 
 install: all
 	$(INSTALL) -m 755 -d $(IROOT)$(etcdir)
