diff -up at-3.1.10/pam_atd.pamkeyring at-3.1.10/pam_atd
--- at-3.1.10/pam_atd.pamkeyring	2008-07-18 16:15:07.000000000 +0200
+++ at-3.1.10/pam_atd	2008-07-18 16:16:48.000000000 +0200
@@ -2,8 +2,8 @@
 #
 #
 auth       sufficient  pam_rootok.so
-auth       include     system-auth
 auth       required    pam_env.so
+auth       include     system-auth
 account        include     system-auth
 session        required    pam_loginuid.so
 session        include     system-auth
