--- at-3.1.10/panic.c.usage	2006-09-07 13:21:23.000000000 +0200
+++ at-3.1.10/panic.c	2006-09-07 13:24:00.000000000 +0200
@@ -92,7 +92,8 @@
 /* Print usage and exit.
  */
     fprintf(stderr, "Usage: at [-V] [-q x] [-f file] [-mldbv] time\n"
-    	    "       at -c job ...\n"
+    	    "       at [-V] [-q x] [-f file] [-m] -t [[CC]YY]MMDDhhmm\n"
+    	    "       at -c job [job...]\n"
 	    "       atq [-V] [-q x]\n"
 	    "       atrm [-V] job ...\n"
 	    "       batch\n");
